experimental/features.py

```diff
@@ -0,0 +1,371 @@
+      W1 = random.normal(rng2, (nngp_feat_shape[-1], feature_dim1))
+      # TensorSRHT of degree 2 for approximating tensor product.
+      ts2 = TensorSRHT2(rng=rng3,
+                        input_dim1=ntk_feat_shape[-1],
+                        input_dim2=feature_dim0,
+                        sketch_dim=sketch_dim).init_sketches()  # pytype:disable=wrong-keyword-args
+      return (new_nngp_feat_shape, new_ntk_feat_shape), (W0, W1, ts2)
+
+    elif method == 'ps':
+      # rng1, rng2, rng3 = random.split(rng, 3)
+      # # PolySketch algorithm for arc-cosine kernel of order 0.
+      # ps0 = PolyTensorSRHT(rng1, nngp_feat_shape[-1], poly_sketch_dim0,
+      #                      poly_degree0)
+      # # PolySketch algorithm for arc-cosine kernel of order 1.
+      # ps1 = PolyTensorSRHT(rng2, nngp_feat_shape[-1], poly_sketch_dim1,
+      #                      poly_degree1)
+      # # TensorSRHT of degree 2 for approximating tensor product.
+      # ts2 = TensorSRHT2(rng3, ntk_feat_shape[-1], feature_dim0, sketch_dim)
+      # return (new_nngp_feat_shape, new_ntk_feat_shape), (ps0, ps1, ts2)
+      raise NotImplementedError
+
+    elif method == 'exact':
+      # The exact feature map computation is for debug.
+      new_nngp_feat_shape = nngp_feat_shape[:-1] + (_prod(
+          nngp_feat_shape[:-1]),)
+      new_ntk_feat_shape = ntk_feat_shape[:-1] + (_prod(ntk_feat_shape[:-1]),)
+      return (new_nngp_feat_shape, new_ntk_feat_shape), ()
+
+  def feature_fn(f: Features, input=None, **kwargs) -> Features:
+
+    input_shape = f.nngp_feat.shape[:-1]
+    nngp_feat_dim = f.nngp_feat.shape[-1]
+    ntk_feat_dim = f.ntk_feat.shape[-1]
+
+    nngp_feat_2d = f.nngp_feat.reshape(-1, nngp_feat_dim)
+    ntk_feat_2d = f.ntk_feat.reshape(-1, ntk_feat_dim)
+
+    if method == 'rf':  # Random Features approach.
+      W0: np.ndarray = input[0]
+      W1: np.ndarray = input[1]
+      ts2: TensorSRHT2 = input[2]
+
+      kappa0_feat = (nngp_feat_2d @ W0 > 0) / np.sqrt(W0.shape[-1])
+      nngp_feat = (np.maximum(nngp_feat_2d @ W1, 0) /
+                   np.sqrt(W1.shape[-1])).reshape(input_shape + (-1,))
+      ntk_feat = ts2.sketch(ntk_feat_2d,
+                            kappa0_feat).reshape(input_shape + (-1,))
+
+    elif method == 'ps':
+      # ps0: PolyTensorSRHT = input[0]
+      # ps1: PolyTensorSRHT = input[1]
+      # ts2: TensorSRHT2 = input[2]
+      raise NotImplementedError
+
+    elif method == 'exact':  # Exact feature extraction via Cholesky decomposition.
+      nngp_feat = cholesky(kappa1(nngp_feat_2d)).reshape(input_shape + (-1,))
+
+      ntk = ntk_feat_2d @ ntk_feat_2d.T
+      kappa0_mat = kappa0(nngp_feat_2d)
+      ntk_feat = cholesky(ntk * kappa0_mat).reshape(input_shape + (-1,))
+
+    else:
+      raise NotImplementedError
+
+    return f.replace(nngp_feat=nngp_feat, ntk_feat=ntk_feat)
+
+  return init_fn, feature_fn
+
+
+def _conv_feat(X, filter_size):
```

Comment:
Could we use the conv-based `jax.lax.conv_general_dilated_patches` here instead of the for-loop? This should make the code much faster and more generic.
https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.conv_general_dilated_patches.html
I'm assuming that this code extracts image patches, lmk if I'm misunderstanding!

Reply:
Thanks! I compared the runtime of `lax.conv_general_dilated_patches` to manual for-loop, but the latter runs much faster. Please check [this colab](https://colab.research.google.com/drive/1YnMPIPNVUHx_N6okBC63tBw85NNJs3iJ?usp=sharing). I would rather keep this implementation for now.

Reply:
Thanks, this is a very interesting result, and I confirm that on Nvidia T4 and on a CPU in public colab your approach is faster! 

When I tried it internally on a V100 (https://colab.research.google.com/gist/icml2022anon/386459c19c064aee3f3b05bb56473c8c/conv_vs_for_loop-nvidia-v100.ipynb) I got the conv-based approach to be slightly faster but not by much (~10%). On a TPU, the conv-based approach is about ~30% faster.

I think we might want to use the conv-based approach eventually (i.e. `concat0` or `concat1` from the colab above) for the purpose of generality (supporting different shapes/strides/padding etc), but happy to keep it as is for now, especially if you do experiments on CPU.